File: case_studies/argo/depth.py
import numpy as np
from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.pyplot as plt
from scipy.interpolate import interp1d
from statsmodels.nonparametric.smoothers_lowess import lowess
import pandas as pd
from read import *

pdf = PdfPages("depth_py.pdf")

ii = np.random.choice(np.arange(temp.shape[1]), 5000, replace=False)
tempx = temp[:, ii]
psalx = psal[:, ii]
latx = lat[ii]
lonx = lon[ii]
dayx = day[ii]

# Calculate the spatial depth of vector v relative
# to all columns of x.
def sdepth(v, x):
    p, n = x.shape
    z = x - v[:, None]
    zn = np.sqrt((z**2).sum(0))
    zn[np.abs(zn) < 1e-12] = np.inf
    z /= zn
    u = z.mean(1)
    return 1 - np.sqrt(np.sum(u**2))

# Calculate the L2 depth of vector v relative
# to all columns of x.
def l2depth(v, x):
    p, n = x.shape
    z = x - v[:, None]
    zn = np.sqrt((z**2).sum(0))
    d = zn.mean()
    return 1e6 / (1 + d)

# Estimate the band depth of vector v relative
# to all columns of x, using 500 random draws
# to estimate the band depth.
def bdepth(v, x, m=500):
    p, n = x.shape
    t = 0.0
    for k in range(m):
        ii = np.random.choice(n, 3, replace=False)
        z = x[:, ii]
        mn = z.min(1)
        mx = z.max(1)
        t += np.mean((v >= mn) & (v <= mx))
    t /= m
    return t

# Calculate the depth of every column of x relative
# to the other columns, using 'dfun' as the depth
# function.
def depths(x, dfun, progress=False):
    p, n = x.shape
    d = np.zeros(n)
    for i in range(n):
        if progress and (i % 200 == 0):
            print(i, end="", flush=True)
            print(".", end="", flush=True)
        d[i] = dfun(x[:, i], x)
    if progress:
        print("done", flush=True)
    return d

# Plot a small random selection of profiles from each depth decile.
def depth_cut(dp, x, q, pressure, ylab):
    dq = pd.qcut(dp, q)
    for (i,iv) in enumerate(dq.categories):
        ii = np.flatnonzero(dq == iv)
        dd = dq[ii]
        jj = np.random.choice(ii, 10)

        plt.clf()
        plt.grid(True)
        plt.title("Depth quantile %d %s" % (i + 1, str(iv)))
        for j in jj:
            plt.plot(pressure, x[:, j], "-", color="grey")
        plt.ylabel(ylab, size=15)
        plt.xlabel("Pressure", size=15)
        pdf.savefig()

dp_temp = depths(tempx, sdepth, progress=True)
dp_psal = depths(psalx, sdepth, progress=True)

q = 10
depth_cut(dp_temp, tempx, q, pressure, "Temperature")
depth_cut(dp_psal, psalx, q, pressure, "Salinity")

# Plot the estimated conditional mean depth relative to each explanatory variable.
# The bands in these plots are +/- f times the mean absolute deviation from the
# conditional mean.
def depth_correlates(dp, lat, lon, day, title, f=2):
    dpx = pd.DataFrame({"depth": dp, "lat": lat, "lon": lon, "day": day})
    vn = {"lat": "Latitude", "lon": "Longitude", "day": "Day"}
    for v in ["lat", "lon", "day"]:
        xx = np.linspace(dpx[v].min(), dpx[v].max(), 100)
        m = lowess(dpx["depth"], dpx[v])
        aresid = np.abs(m[:, 1] - dpx["depth"])
        r = lowess(aresid, dpx[v])
        dh = interp1d(m[:, 0], m[:, 1])(xx)
        dq = interp1d(r[:, 0], r[:, 1])(xx)
        plt.clf()
        plt.title(title)
        plt.grid(True)
        plt.plot(xx, dh, "-")
        plt.fill_between(xx, dh-f*dq, dh+f*dq, color="grey", alpha=0.5)
        plt.xlabel(vn[v], size=15)
        plt.ylabel("Depth", size=15)
        pdf.savefig()

depth_correlates(dp_temp, latx, lonx, dayx, "Atlantic ocean temperature")
depth_correlates(dp_psal, latx, lonx, dayx, "Atlantic ocean salinity")

# Northern hemisphere
ii = np.flatnonzero(latx > 0)
depth_correlates(dp_temp[ii], latx[ii], lonx[ii], dayx[ii], "Northern hemisphere temperature")
depth_correlates(dp_psal[ii], latx[ii], lonx[ii], dayx[ii], "Northern hemisphere salinity")

# Southern hemisphere
ii = np.flatnonzero(latx < 0)
depth_correlates(dp_temp[ii], latx[ii], lonx[ii], dayx[ii], "Southern hemisphere temperature")
depth_correlates(dp_psal[ii], latx[ii], lonx[ii], dayx[ii], "Southern hemisphere salinity")

pdf.close()

File: case_studies/argo/depth.R
library(dplyr)
library(ggplot2)

source("read.R")

# The code below takes awhile to run so subsample columns.
m = 10000
ii = sample(1:dim(temp)[2], m, replace=F)
tempx = temp[,ii]
psalx = psal[,ii]
latx = lat[ii]
lonx = lon[ii]
dayx = day[ii]

# Calculate the spatial depth of column i of x relative
# to the other columns of x.
sdepth = function(i, x) {
    p = dim(x)[1]
    n = dim(x)[2]
    z = x - x[,i]
    zn = sqrt(colSums(z^2))
    zn[i] = Inf
    z = z / outer(array(1, p), zn)
    u = apply(z, 1, mean)
    return(1 - sqrt(sum(u^2)))
}

# Calculate the L2 depth of column i of x relative
# to the other columns of x.
l2depth = function(i, x) {
    p = dim(x)[1]
    n = dim(x)[2]
    z = x - x[,i]
    zn = sqrt(colSums(z^2))
    return(1e6 / (1 + mean(zn)))
}

# Calculate the depth of every column of x relative
# to the other columns, using the given depth function.
depths = function(x, dfun, progress=F) {
    n = dim(x)[2]
    d = array(0, n)
    for (i in 1:n) {
        if (progress && (i %% 200 == 0)) {
            cat(i)
            cat(".")
        }
        d[i] = dfun(i, x)
    }
    if (progress) {
        cat("\n")
    }
    return(d)
}

pdf("depth_r.pdf")

plot_mean = function(x, pressure, ylab) {
    y = apply(x, 1, mean)
    da = data.frame(pressure=pressure, y=y)
    plt = ggplot(aes(x=pressure, y=y), data=da) + geom_line()
    plt = plt + ylab(ylab) + xlab("Pressure")
    print(plt)
}

# Plot the mean curves for temperature and salinity.
plot_mean(temp, pressure, "Mean temperature")
plot_mean(psal, pressure, "Mean salinity")

# Partition the depths into q equal-sized bins.  Within each
# bin plot the response variable (temperature or salinity)
# against pressure for 10 randomly selected profiles.
depth_cut = function(x, q, pressure, ylab) {
    dp = depths(x, l2depth, progress=T)
    dq = ntile(dp, q)

    for (i in seq(1, q, 2)) {

        da = data.frame()
        for (k in c(i, i+1)) {
            ii = which(dq == k)
            jj = sample(ii, 10)
            for (j in jj) {
                dd = data.frame(pressure=pressure, y=x[,j], j=j, k=k)
                da = rbind(da, dd)
            }
        }
        da$j = as.factor(da$j)
        da$k = as.factor(da$k)

        plt = ggplot(aes(x=pressure, y=y, color=k, group=j), data=da) + geom_line()
        plt = plt + ggtitle(sprintf("Depth band %d-%d", i, i+1))
        plt = plt + ylab(ylab) + xlab("Pressure")
        print(plt)
    }
    return(dp)
}

# Plot a small random selection of profiles from each depth interval.
q = 20
dp_temp = depth_cut(tempx, q, pressure, "Temperature")
dp_psal = depth_cut(psalx, q, pressure, "Salinity")

# Plot the estimated conditional mean depth relative to each explanatory variable.
# The bands in these plots are +/- 2 times the mean absolute deviation from the
# conditional mean.
depth_correlates = function(dp, ttl) {
    dpx = data.frame(depth=dp, lat=latx, lon=lonx, day=dayx)
    for (v in c("lat", "lon", "day")) {
        xx = seq(min(dpx[[v]]), max(dpx[[v]]), length.out=100)
        m = lowess(dpx[[v]], dpx$depth)
        aresid = abs(m$y - dpx$depth)
        r = lowess(dpx[[v]], aresid)
        dh = approxfun(m$x, m$y)(xx)
        dq = approxfun(r$x, r$y)(xx)
        da = data.frame(x=xx, depth=dh, r=dq)
        f = 2
        da$ymin = dh - f*dq
        da$ymax = dh + f*dq
        plt = ggplot(aes(x=x, y=depth), data=da)
        plt = plt + geom_ribbon(aes(x=x, ymin=ymin, ymax=ymax, y=depth), fill="grey70", data=da) + geom_line()
        plt = plt + labs(x=v, y="Depth") + ggtitle(ttl)
        print(plt)
    }
}

depth_correlates(dp_temp, "Temperature")
depth_correlates(dp_psal, "Salinity")

dev.off()
